### PR TITLE
Gaussian noise voxel support on HOST and HIP

### PR DESCRIPTION
**Pull request by sampath1117**
- Adds support for U8, F32 datatypes
- Adds support in test suite

### CODE REVIEW

File: utilities/test_suite/HOST/Tensor_voxel_host.cpp [hunk 1]
```diff
@@ -251,6 +251,23 @@ int main(int argc, char * argv[])
...
+                    Rpp32f stdDevTensor[batchSize];
+                    Rpp32u seed = 1255459;
+                    for (int i = 0; i < batchSize; i++)
+                    {
+                        meanTensor[i] = 10.0f;
+                        stdDevTensor[i] = 5.0f;
+                    }
+
+                    startWallTime = omp_get_wtime();
+                    rppt_gaussian_noise_3d_host(inputF32, descriptorPtr3D, outputF32, descriptorPtr3D, meanTensor, stdDevTensor, seed, roiGenericSrcPtr, roiTypeSrc, handle);
```
Comment (r-abishek) [inline utilities/test_suite/HOST/Tensor_voxel_host.cpp:268]: Have we added U8/F16/F32/I8 on host? But tests here are only for F32?

File: src/modules/cpu/kernel/noise_gaussian.hpp [hunk 2]
```diff
@@ -1390,3 +1390,752 @@ RppStatus gaussian_noise_i8_i8_host_tensor(Rpp8s *srcPtr,
...
+
+                    int vectorLoopCount = 0;
+                    for (; vectorLoopCount < alignedLength; vectorLoopCount += vectorIncrement)
+                    {
+#if __AVX2__
+                        __m256 p[6];
+                        rpp_simd_load(rpp_load48_u8pkd3_to_f32pln3_avx, srcPtrTemp, p);                                 // simd loads
+                        rpp_multiply48_constant(p, avx_p1op255);                                                        // u8 normalization to range[0,1]
+                        compute_gaussian_noise_48_host(p, pxXorwowStateX, &pxXorwowStateCounter, pGaussianNoiseParams); // gaussian_noise adjustment
+                        rpp_multiply48_constant(p, avx_p255);                                                           // u8 un-normalization
```
Comment (r-abishek) [inline src/modules/cpu/kernel/noise_gaussian.hpp:1480]: Not sure if the adjustment for 255 is needed in a tensor. Lets discuss and come to consensus for this.

File: docs/sphinx/requirements.in [hunk 3]
```diff
@@ -1 +1 @@
-rocm-docs-core[api_reference]==0.34.0
```
Comment (r-abishek) [inline docs/sphinx/requirements.in:1]: These shouldn't change in the PR. Pls revert

File: docs/sphinx/requirements.txt [hunk 4]
```diff
@@ -112,7 +112,7 @@ requests==2.28.2
     # via
     #   pygithub
     #   sphinx
-rocm-docs-core[api-reference]==0.34.0
+rocm-docs-core[api-reference]==0.34.2
```
Comment (r-abishek) [inline docs/sphinx/requirements.txt:115]: Pls revert this too

File: include/rppt_tensor_arithmetic_operations.h [hunk 5]
```diff
@@ -160,10 +168,92 @@ RppStatus rppt_subtract_scalar_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGener
...
+ * \param[in] dstGenericDescPtr destination tensor descriptor
+ * \param[in] mulTensor multiplier values for used for multiplication (1D tensor of batchSize Rpp32f values)
+ * \param[in] roiGenericPtrSrc ROI data for each image in source tensor (tensor of batchSize RpptRoiGeneric values)
+ * \param[in] roiType ROI type used (RpptRoi3DType::XYZWHD or RpptRoi3DType::LTFRBB)
+ * \param [in] rppHandle RPP HOST handle created with <tt>\ref rppCreateWithBatchSize()</tt>
+ * \return A <tt> \ref RppStatus</tt> enumeration.
+ * \retval RPP_SUCCESS Successful completion.
+ * \retval RPP_ERROR* Unsuccessful completion.
+ */
+RppStatus rppt_multiply_scalar_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDescPtr, RppPtr_t dstPtr, RpptGenericDescPtr dstGenericDescPtr, Rpp32f *subtractTensor, RpptROI3DPtr roiGenericPtrSrc, RpptRoi3DType roiType, rppHandle_t rppHandle);
```
Comment (r-abishek) [inline include/rppt_tensor_arithmetic_operations.h:187]: Pls sync this branch

**Review by r-abishek — CHANGES_REQUESTED** [review]
@sampath1117 This PR needs to sync with upstream develop

**Comment by sampath1117** [comment]
@r-abishek 
The branch is already in sync with develop. I think there is some issue in github UI. I just changed the base to master and rechanged to develop

Now it is showing correct file changes

**Comment by sampath1117** [comment]
@r-abishek 
I have merged the PR branch with latest changes in opensource develop and resolved the conflicts

Please update your develop branch so that merge conflicts that are showing currently will be not shown

File: include/rppt_tensor_effects_augmentations.h [hunk 6]
```diff
@@ -418,6 +418,45 @@ RppStatus rppt_ricap_host(RppPtr_t srcPtr, RpptDescPtr srcDescPtr, RppPtr_t dstP
 RppStatus rppt_ricap_gpu(RppPtr_t srcPtr, RpptDescPtr srcDescPtr, RppPtr_t dstPtr, RpptDescPtr dstDescPtr, Rpp32u *permutedIndicesTensor, RpptROIPtr roiPtrInputCropRegion, RpptRoiType roiType, rppHandle_t rppHandle);
 #endif // GPU_SUPPORT
 
+/*! \brief  Gaussian noise augmentation on HOST backend
+ * \details This function adds gaussian noise to a batch of 4D tensors.
```
Comment (r-abishek) [inline include/rppt_tensor_effects_augmentations.h:422]: We say "a batch of 4D" here, but the name of the API is "_3d". That might create some confusion.
So please extend the comment a bit to make it crystal clear.
Open to solutions, but possibly explain something like (A batch of N 3D voxel tensors - each XYZ voxel coordinate supporting 1 or 3 channels)

File: src/modules/rppt_tensor_effects_augmentations.cpp [hunk 7]
```diff
@@ -434,6 +434,65 @@ RppStatus rppt_gaussian_noise_host(RppPtr_t srcPtr,
...
+    if ((srcGenericDescPtr->layout != RpptLayout::NCDHW) && (srcGenericDescPtr->layout != RpptLayout::NDHWC)) return RPP_ERROR_INVALID_SRC_LAYOUT;
+    if ((dstGenericDescPtr->layout != RpptLayout::NCDHW) && (dstGenericDescPtr->layout != RpptLayout::NDHWC)) return RPP_ERROR_INVALID_DST_LAYOUT;
+    if (srcGenericDescPtr->layout != dstGenericDescPtr->layout) return RPP_ERROR_INVALID_ARGUMENTS;
+
+    RpptXorwowStateBoxMuller xorwowInitialState[SIMD_FLOAT_VECTOR_LENGTH];
+    rpp_host_rng_xorwow_f32_initialize_multiseed_stream_boxmuller<SIMD_FLOAT_VECTOR_LENGTH>(xorwowInitialState, seed);
+
+    if ((srcGenericDescPtr->dataType == RpptDataType::U8) && (dstGenericDescPtr->dataType == RpptDataType::U8))
+    {
+        gaussian_noise_3d_u8_u8_host_tensor(static_cast<Rpp8u*>(srcPtr) + srcGenericDescPtr->offsetInBytes,
```
Comment (r-abishek) [inline src/modules/rppt_tensor_effects_augmentations.cpp:462]: How come we support u8, f32 in HOST, but just f32 in HIP?

File: src/modules/hip/kernel/noise_gaussian.hpp [hunk 8]
```diff
@@ -253,6 +273,102 @@ __global__ void gaussian_noise_pln3_pkd3_hip_tensor(T *srcPtr,
...
+    xorwowState.x[0] = xorwowInitialStatePtr->x[0] + xorwowSeed;
+    xorwowState.x[1] = xorwowInitialStatePtr->x[1] + xorwowSeed;
+    xorwowState.x[2] = xorwowInitialStatePtr->x[2] + xorwowSeed;
+    xorwowState.x[3] = xorwowInitialStatePtr->x[3] + xorwowSeed;
+    xorwowState.x[4] = xorwowInitialStatePtr->x[4] + xorwowSeed;
+    xorwowState.counter = xorwowInitialStatePtr->counter + xorwowSeed;
+
+    d_float24 val_f24;
+    rpp_hip_load24_pkd3_and_unpack_to_float24_pln3(srcPtr + srcIdx, &val_f24);
+    bool copyInput = (!gaussianNoise3dParams_f2.x) && (!gaussianNoise3dParams_f2.y);
```
Comment (r-abishek) [inline src/modules/hip/kernel/noise_gaussian.hpp:366]: When does gaussian noise give output = input? Isn't it for mean = 0 and stdDev = 1? Pls check condition

File: src/modules/hip/kernel/noise_gaussian.hpp [hunk 9]
```diff
@@ -351,3 +467,68 @@ RppStatus hip_exec_gaussian_noise_tensor(T *srcPtr,
...
+                               dim3(ceil((float)globalThreads_x/LOCAL_THREADS_X), ceil((float)globalThreads_y/LOCAL_THREADS_Y), ceil((float)globalThreads_z/LOCAL_THREADS_Z)),
+                               dim3(LOCAL_THREADS_X, LOCAL_THREADS_Y, LOCAL_THREADS_Z),
+                               0,
+                               handle.GetStream(),
+                               srcPtr + (batchCount * srcGenericDescPtr->strides[0]),
+                               make_uint3(srcGenericDescPtr->strides[1], srcGenericDescPtr->strides[2], srcGenericDescPtr->strides[3]),
+                               dstPtr + (batchCount * dstGenericDescPtr->strides[0]),
+                               make_uint3(dstGenericDescPtr->strides[1], dstGenericDescPtr->strides[2], dstGenericDescPtr->strides[3]),
+                               dstGenericDescPtr->dims[1],
+                               make_float2(meanTensor[batchCount], stdDevTensor[batchCount]),
```
Comment (r-abishek) [inline src/modules/hip/kernel/noise_gaussian.hpp:503]: Have we tried doing a simple hipMemcpy inside the loop if meanTensor[batchCount] and stdDevTensor[batchCount] condition is satisfied?
If not, we call the kernel to add noise.

**Review by r-abishek — CHANGES_REQUESTED** [review]
@sampath1117 Please address comments

File: include/rppt_tensor_effects_augmentations.h (hunk 6 above)
Comment (sampath1117) [reply include/rppt_tensor_effects_augmentations.h:422]: We can this to voxel, which is the suffix we used for flip too, flip_voxel
The gaussian noise HIP API initially had suffix 3d, so i thought i might need to retain 3D for HOST itself and should not change to any other name

Will change the name to voxel

File: src/modules/rppt_tensor_effects_augmentations.cpp (hunk 7 above)
Comment (sampath1117) [reply src/modules/rppt_tensor_effects_augmentations.cpp]: Modified HIP outer api to add support for U8

**Review by r-abishek — APPROVED** [review]

**Pull request merged**